File: plugin/src/notification-envelope.js
// Notification envelope v1 (ADR 0008).
//
// The encrypted Agent Notification payload carried from the notify hook
// through the Push Relay and APNs to the app's service extension. Wire
// format: a compact JSON object `{"v":1,"kid":...,"n":...,"ct":...}` whose
// `ct` is the AES-256-GCM ciphertext (plus tag) of a compact JSON plaintext
// `{"pane":...,"kind":...,"status":...,"ts":...}` plus the optional display
// fields `project` and `title`. See README.md for the
// schema and test-vectors/notification-payload-v1.json for the
// cross-implementation vectors: this side proves the encrypt direction, the
// Swift side proves the decrypt direction. Breaking changes bump the
// version, which both implementations must honor together.

import { createCipheriv, createHash, randomBytes } from "node:crypto";

export const NOTIFICATION_ENVELOPE_VERSION = 1;

// Additional authenticated data binding the ciphertext to envelope v1, so a
// re-framed ciphertext under a different version fails authentication.
const AAD = Buffer.from(`HERDR-NOTIFY:${NOTIFICATION_ENVELOPE_VERSION}`, "utf8");

const KEY_BYTES = 32;
const NONCE_BYTES = 12;
const KEY_ID_BYTES = 8;

export class NotificationEnvelopeError extends Error {
  /**
   * @param {"bad_envelope"|"unsupported_version"|"decrypt_failed"|"bad_payload"} code
   * @param {string} message
   */
  constructor(code, message) {
    super(message);
    this.name = "NotificationEnvelopeError";
    this.code = code;
  }
}

function fail(code, message) {
  throw new NotificationEnvelopeError(code, message);
}

/**
 * Derive the cleartext key id for a raw 32-byte Notification Key: the first
 * 8 bytes of SHA-256 over the key, unpadded base64url. Both ends derive it,
 * so it never needs to be stored or exchanged separately.
 *
 * @param {Buffer} key
 * @returns {string}
 */
export function notificationKeyId(key) {
  if (!Buffer.isBuffer(key) || key.length !== KEY_BYTES) {
    throw new TypeError(`Notification Key must be a ${KEY_BYTES}-byte Buffer`);
  }
  return createHash("sha256").update(key).digest().subarray(0, KEY_ID_BYTES).toString("base64url");
}

// Hard ceiling for the display-only strings. The notify hook trims to
// something shorter still (see DISPLAY_LIMIT there); this is the contract's
// backstop keeping an APNs payload well inside its 4 KB budget once the
// ciphertext is base64url'd.
const DISPLAY_FIELD_MAX = 256;

/**
 * Validate the notification payload shape. `project` and `title` are the
 * optional display fields: omit them (or pass null) and the app renders one
 * step less specific.
 *
 * @param {object} payload
 * @param {string} payload.paneId herdr pane id the Agent lives in
 * @param {string} payload.agentKind agent kind as herdr reports it
 * @param {string} payload.status the new Agent Status (lenient open set)
 * @param {number} payload.timestamp unix-seconds of the status transition
 * @param {string|null} [payload.project] workspace label the Agent runs in
 * @param {string|null} [payload.title] Agent terminal title, glyphs stripped
 */
function validatePayload(payload) {
  const { paneId, agentKind, status, timestamp } = payload;
  if (typeof paneId !== "string" || paneId.length === 0) {
    fail("bad_payload", "paneId must be a non-empty string");
  }
  if (typeof agentKind !== "string" || agentKind.length === 0) {
    fail("bad_payload", "agentKind must be a non-empty string");
  }
  if (typeof status !== "string" || status.length === 0) {
    fail("bad_payload", "status must be a non-empty string");
  }
  if (!Number.isInteger(timestamp) || timestamp <= 0) {
    fail("bad_payload", `timestamp must be a positive unix-seconds integer, got ${timestamp}`);
  }
  for (const field of ["project", "title"]) {
    const value = payload[field];
    if (value === undefined || value === null) continue;
    if (typeof value !== "string") fail("bad_payload", `${field} must be a string when present`);
    if (value.length > DISPLAY_FIELD_MAX) {
      fail("bad_payload", `${field} must be at most ${DISPLAY_FIELD_MAX} characters`);
    }
  }
}

/**
 * Encrypt a notification payload into its canonical v1 envelope string.
 *
 * Key order in both JSON objects is fixed so the same inputs always yield
 * the same envelope; the shared vectors assert on the exact string.
 *
 * @param {object} payload see {@link validatePayload}
 * @param {Buffer} key raw 32-byte Notification Key
 * @param {{nonce?: Buffer}} [options] injectable 12-byte nonce (tests only;
 *   production must use the random default)
 * @returns {string} the canonical envelope JSON string
 * @throws {NotificationEnvelopeError} with code `bad_payload`
 */
export function encryptNotificationEnvelope(payload, key, { nonce } = {}) {
  const keyId = notificationKeyId(key);
  if (nonce === undefined) {
    nonce = randomBytes(NONCE_BYTES);
  } else if (!Buffer.isBuffer(nonce) || nonce.length !== NONCE_BYTES) {
    throw new TypeError(`nonce must be a ${NONCE_BYTES}-byte Buffer`);
  }
  validatePayload(payload);

  // Empty display fields are omitted rather than sent blank, so the encoding
  // stays canonical whichever way the Host failed to resolve them.
  const plaintext = JSON.stringify({
    pane: payload.paneId,
    kind: payload.agentKind,
    status: payload.status,
    ts: payload.timestamp,
    ...(payload.project ? { project: payload.project } : {}),
    ...(payload.title ? { title: payload.title } : {}),
  });
  const cipher = createCipheriv("aes-256-gcm", key, nonce);
  cipher.setAAD(AAD);
  const ciphertext = Buffer.concat([
    cipher.update(plaintext, "utf8"),
    cipher.final(),
    cipher.getAuthTag(),
  ]);

  return JSON.stringify({
    v: NOTIFICATION_ENVELOPE_VERSION,
    kid: keyId,
    n: nonce.toString("base64url"),
    ct: ciphertext.toString("base64url"),
  });
}

File: Sources/HeelerNotificationCore/NotificationEnvelope.swift
import Foundation

/// A decrypted Agent Notification payload: what a Host's notify hook tells
/// this device about an Agent Status transition (ADR 0008).
struct NotificationPayload: Sendable, Equatable {
    /// The herdr pane id the Agent lives in; drives the Attach deep link.
    let paneID: String
    /// The agent kind as herdr reports it (claude, codex, ...).
    let agentKind: String
    /// The new Agent Status. An open set like everywhere else on the herdr
    /// wire: unrecognized values pass through for forward compatibility.
    let status: AgentStatus
    /// When the transition happened.
    let timestamp: Date
    /// The workspace label the Agent runs in — the project name the alert
    /// leads with. Nil when the Host could not resolve it, or when the plugin
    /// predates the field.
    let project: String?
    /// The Agent's terminal title, stripped of status glyphs: what the agent
    /// is working on. Nil under the same conditions as `project`.
    let title: String?

    init(
        paneID: String, agentKind: String, status: AgentStatus, timestamp: Date,
        project: String? = nil, title: String? = nil
    ) {
        self.paneID = paneID
        self.agentKind = agentKind
        self.status = status
        self.timestamp = timestamp
        self.project = project
        self.title = title
    }
}

/// The notification envelope v1: the encrypted payload carried from the
/// plugin's notify hook through the Push Relay and APNs to this app
/// (ADR 0008). Wire format, a compact JSON object:
///
///     {"v": 1, "kid": "...", "n": "...", "ct": "..."}
///
/// `ct` is the AES-256-GCM ciphertext (plus 16-byte tag) of a compact JSON
/// plaintext `{"pane", "kind", "status", "ts"}` under the per-host
/// Notification Key, with the version string bound as additional
/// authenticated data. The schema lives in `plugin/README.md`; the shared
/// test vectors in `plugin/test-vectors/notification-payload-v1.json` are
/// the single source of truth for this decoder and the plugin's encoder
/// (which proves the encrypt direction). Unknown fields at either layer are
/// ignored (additive v1 metadata); breaking changes bump the version, which
/// both implementations must honor together.
enum NotificationEnvelope {
    static let version = 1

    /// Additional authenticated data binding the ciphertext to envelope v1,
    /// so a re-framed ciphertext under a different version fails to open.
    private static var additionalData: Data {
        Data("HERDR-NOTIFY:\(version)".utf8)
    }

    /// Derives the cleartext key id for a raw 32-byte Notification Key: the
    /// first 8 bytes of SHA-256 over the key, unpadded base64url. Both ends
    /// derive it, so it never needs to be stored or exchanged separately;
    /// the app uses it to select the right key (and thus Host) when several
    /// Hosts are registered.
    static func keyID(for key: Data) -> String {
        SealedEnvelopeCodec.keyID(for: key)
    }

    /// Reads the cleartext key id off an envelope without decrypting it, so
    /// the right Notification Key can be selected when several Hosts are
    /// registered. Nil when the framing is too broken to carry one; version
    /// and ciphertext checks stay `decrypt`'s job.
    static func peekKeyID(in envelope: Data) -> String? {
        SealedEnvelopeCodec.peekKeyID(in: envelope)
    }

    /// The shared front half of alert rendering (#71) and tap routing (#74):
    /// read the envelope string off a push's `userInfo`, select the
    /// Notification Key record by the envelope's kid, and decrypt. Nil for
    /// anything undecryptable — missing or non-string envelope, unknown kid,
    /// any `NotificationEnvelopeError` — and each caller picks its own
    /// fallback (generic banner, Console).
    static func open(
        userInfo: [AnyHashable: Any], keys: [NotificationKeyRecord]
    ) -> (record: NotificationKeyRecord, payload: NotificationPayload)? {
        guard let envelopeText = userInfo["envelope"] as? String else { return nil }
        let envelope = Data(envelopeText.utf8)
        guard let kid = peekKeyID(in: envelope),
            let record = keys.first(where: { $0.keyID == kid }),
            let payload = try? decrypt(envelope, using: record.key)
        else { return nil }
        return (record, payload)
    }

    /// Decrypts an envelope's wire bytes with a Notification Key.
    ///
    /// Every way an envelope can be undecryptable — malformed framing, a
    /// future version, tampered or wrongly keyed ciphertext, a garbage
    /// plaintext — surfaces as a typed `NotificationEnvelopeError` so the
    /// service extension can fall back to a generic banner; nothing here
    /// crashes on hostile input.
    static func decrypt(
        _ envelope: Data, using key: Data
    ) throws(NotificationEnvelopeError) -> NotificationPayload {
        let plaintext = try SealedEnvelopeCodec.open(
            envelope, using: key, authenticating: additionalData)
        return try validated(plaintext)
    }

    private static func validated(
        _ plaintext: Data
    ) throws(NotificationEnvelopeError) -> NotificationPayload {
        let wire: WirePlaintext
        do {
            wire = try JSONDecoder().decode(WirePlaintext.self, from: plaintext)
        } catch {
            throw .badPayload(reason: "plaintext is not a well-formed JSON object")
        }
        guard let pane = wire.pane, !pane.isEmpty else {
            throw .badPayload(reason: "pane must be a non-empty string")
        }
        guard let kind = wire.kind, !kind.isEmpty else {
            throw .badPayload(reason: "kind must be a non-empty string")
        }
        guard let status = wire.status, !status.isEmpty else {
            throw .badPayload(reason: "status must be a non-empty string")
        }
        guard let rawTimestamp = wire.ts, let timestamp = Int(exactly: rawTimestamp),
            timestamp > 0
        else {
            throw .badPayload(reason: "ts must be a positive unix-seconds integer")
        }
        // Display-only and additive: a payload without them still renders,
        // one step less specific. Empty strings mean "the Host had nothing",
        // which is the same thing as absent.
        return NotificationPayload(
            paneID: pane, agentKind: kind, status: AgentStatus(rawValue: status),
            timestamp: Date(timeIntervalSince1970: TimeInterval(timestamp)),
            project: nonEmpty(wire.project), title: nonEmpty(wire.title))
    }

    private static func nonEmpty(_ text: String?) -> String? {
        guard let text, !text.isEmpty else { return nil }
        return text
    }

    /// Decrypted JSON wire shape, lenient for the same reason. `project` and
    /// `title` are the additive v1 display fields: absent from older plugins,
    /// and never load-bearing for anything but copy.
    private struct WirePlaintext: Decodable {
        var pane: String?
        var kind: String?
        var status: String?
        var ts: Double?
        var project: String?
        var title: String?
    }
}

/// Why an envelope did not yield a payload. These map 1:1 to the error
/// identifiers in the shared test vectors. Any of them means the service
/// extension shows the generic fallback banner instead of decrypted content.
enum NotificationEnvelopeError: Error, Sendable, Equatable {
    /// The cleartext framing is broken: not a JSON object, or a missing or
    /// mistyped field.
    case badEnvelope(reason: String)
    /// An envelope from a plugin speaking another contract version.
    case unsupportedVersion(found: Int)
    /// GCM authentication failed: tampered material or the wrong key.
    case decryptFailed
    /// The ciphertext opened but its plaintext violates the payload schema.
    case badPayload(reason: String)

    /// The cross-implementation identifier used by the shared test vectors.
    var wireCode: String {
        switch self {
        case .badEnvelope: "bad_envelope"
        case .unsupportedVersion: "unsupported_version"
        case .decryptFailed: "decrypt_failed"
        case .badPayload: "bad_payload"
        }
    }
}
